### DISCUSSION TITLE
Fading light with gamma correction has inconsistent speed and continues briefly after stop command

### DISCUSSION THREAD

**Discussion by jonnevw** [arendst/Tasmota#17609] [Ideas]
### PROBLEM DESCRIPTION
- Fading with gamma correction enabled works incorrectly: when setting speed 40 you'd expect a fade from 0 to 10% to take 2000ms (10% x 40 x 500ms), instead it takes about 300ms
- Stopping an in progress fade with 'dimmer !' causes the fade to continue for a small while sometimes in the opposite direction, eg. if dimmer is set to 0, followed by dimmer 100 that is interrupted quickly by a dimmer ! command sometimes causes the increasing fade to decrease for a short period

After debugging the source code the issue has been found to be caused by a mismatch between normal & fast gamma tables, see details further down below

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Shelly RGBW2
- [ ] Tasmota binary firmware version number used: 12.3.1.1
  - [x] Pre-compiled - first identified the issue in official release
  - [x] Self-compiled - reconfirmed the issue in latest dev version
- [x] Flashing tools used: web upload
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
19:36:30.158 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 26, Pld -99, Data 'Template; Module; GPIO 255'
19:36:30.234 SRC: Backlog
19:36:30.237 CMD: Grp 0, Cmd 'TEMPLATE', Idx 1, Len 0, Pld -99, Data ''
19:36:30.243 MQT: stat/tasmota_D96076/RESULT = {"NAME":"Shelly RGBW2","GPIO":[0,0,288,0,0,1,0,0,416,32,0,417,0,0],"FLAG":0,"BASE":18}
19:36:30.450 SRC: Backlog
19:36:30.453 CMD: Grp 0, Cmd 'MODULE', Idx 1, Len 0, Pld -99, Data ''
19:36:30.458 MQT: stat/tasmota_D96076/RESULT = {"Module":{"0":"Shelly RGBW2"}}
19:36:30.656 SRC: Backlog
19:36:30.658 CMD: Grp 0, Cmd 'GPIO', Idx 1, Len 3, Pld 255, Data '255'
19:36:30.668 MQT: stat/tasmota_D96076/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"288":"Led1"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"416":"PWM1"},"GPIO13":{"32":"Button1"},"GPIO14":{"0":"None"},"GPIO15":{"417":"PWM2"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
19:37:33.245 MQT: stat/tasmota_D96076/STATUS = {"Status":{"Module":0,"DeviceName":"ShellyRGBW2Tasmota","FriendlyName":["Shelly RGBW2 Tasmota"],"Topic":"tasmota_D96076","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[1,11,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
19:37:33.260 MQT: stat/tasmota_D96076/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"External System","Uptime":"0T00:45:45","StartupUTC":"2022-12-21T17:51:48","Sleep":50,"CfgHolder":4617,"BootCount":51,"BCResetTime":"2022-12-18T09:43:33","SaveCount":789,"SaveAddress":"F7000"}}
19:37:33.274 MQT: stat/tasmota_D96076/STATUS2 = {"StatusFWR":{"Version":"12.3.1.1(tasmota)","BuildDateTime":"2022-12-21T16:35:25","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"387/699"}}
19:37:33.288 MQT: stat/tasmota_D96076/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":4,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["WIFIHS-5G",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","0A05C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
19:37:33.308 MQT: stat/tasmota_D96076/STATUS4 = {"StatusMEM":{"ProgramSize":636,"Free":364,"Heap":20,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"156085","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
19:37:33.323 MQT: stat/tasmota_D96076/STATUS5 = {"StatusNET":{"Hostname":"tasmota-D96076-0118","IPAddress":"192.168.1.44","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"58:BF:25:D9:60:76","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
19:37:33.333 MQT: stat/tasmota_D96076/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.1.6","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_D96076","MqttUser":"xxxx","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
19:37:33.347 MQT: stat/tasmota_D96076/STATUS7 = {"StatusTIM":{"UTC":"2022-12-21T18:37:33","Local":"2022-12-21T19:37:33","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"08:41","Sunset":"16:55"}}
19:37:33.360 MQT: stat/tasmota_D96076/STATUS10 = {"StatusSNS":{"Time":"2022-12-21T19:37:33"}}
19:37:33.372 MQT: stat/tasmota_D96076/STATUS11 = {"StatusSTS":{"Time":"2022-12-21T19:37:33","Uptime":"0T00:45:45","UptimeSec":2745,"Heap":22,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"POWER":"ON","Dimmer":25,"Color":"4000","White":25,"CT":153,"Channel":[25,0],"Fade":"ON","Speed":4,"LedTable":"ON","Wifi":{"AP":1,"SSId":"xxxx","BSSId":"54:83:3A:78:02:EA","Channel":6,"Mode":"11n","RSSI":86,"Signal":-57,"LinkCount":1,"Downtime":"0T00:00:03"}}}


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
<not included as there is limited logging on fading by default. I added my own loggings in my local version and can share if considered helpful>

```

### TO REPRODUCE
_Steps to reproduce the behavior:_
A. backlog ledtable 1; speed 40; dimmer 0; dimmer 10;
B. backlog ledtable 1; speed 40; dimmer 0; dimmer 100; --> wait for a second --> dimmer !

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

A. the fade from 0 to 10 takes about 300ms while it should take 2000ms
B. after issueing the dimmer ! command you'd expect the dimming to stop immediately. Instead the dimming stops and sometimes continues for a brief moment in the opposite direction

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

After debugging the source code the issue was pinpointed to be in the normal & fast gamma tables:

in xdrv_04_light_utils.ino tables are defined as:
```lua
const gamma_table_t gamma_table[] = {   // don't put in PROGMEM for performance reasons
  {    1,      1 },
  {    4,      1 },
  {  209,     13 },
  {  312,     41 },
  {  457,    106 },
  {  626,    261 },
  {  762,    450 },
  {  895,    703 },
  { 1023,   1023 },
  { 0xFFFF, 0xFFFF }          // fail-safe if out of range
};

// simplified Gamma table for Fade, cheating a little at low brightness
const gamma_table_t gamma_table_fast[] = {
  {   384,    192 },
  {   768,    576 },
  {  1023,   1023 },
  { 0xFFFF, 0xFFFF }          // fail-safe if out of range
};
```

comparing these 2 it is clear they reflect a very different curve, eg. at color ~765 the gamma correction in the normal curve is ~450, while in the fast curve it is ~576.

this is causing both issues:
- for the dimming speed at low brightness first the 10% brightness is gamma corrected using the normal gamma table. Subsequently the duration of the fade is calculated in the LightApplyFade function using the fadeGammaReverse function using the fast gamma table leading to a too low result, causing the duration to be ~1% of the max fade time instead of the expected 10%. 
- when an in progress fade is stopped using the dimmer ! command in the function LightGetCurFadeBri fadeGammaReverse(i, Light.fade_cur_10[i]) is used to calculate back the brightness % using the fast gamma curve table, which is than set as the new brightness. In the subsequent LightAnimate function call the new brightness is translated to a new gamma corrected color using the normal gamma table, which due to the mismatch in gamma tables results in a color that can be significantly higher or lower than the current color, triggering another fade cycle causing the brightness to change up or down for a moment right after the dimmer ! command

As a first solution I tried to solve the problem by aligning the fast gamma table closer to the normal gamma table, eg.
```lua
// simplified Gamma table for Fade, cheating a little at low brightness
const gamma_table_t gamma_table_fast[] = {
//  {   384,    192 },
//  {   768,    576 },
  {   457,    96  },
  {   762,    450 },
  {  1023,   1023 },
  { 0xFFFF, 0xFFFF }          // fail-safe if out of range
};
```
This solves the first timing issue; while there are still small timing inconsistencies across the curve, this is hardly noticable. While the second issue improves, the effect remains visible especially at low brightness. Therefore a better solution would be to use a single gamma table (probably best to take the short fast table to keep performance while accepting that the gamma curve will be slightly less ideal).

Additionally there is a small issue in the LightApplyFade function:
```lua

  uint16_t fade_current = now - Light.fade_start;   // number of milliseconds since start of fade
  if (fade_current <= Light.fade_duration) {    // fade not finished
    for (uint32_t i = 0; i < Light.subtype; i++) {
      Light.fade_cur_10[i] = fadeGamma(i,
                                changeUIntScale(fadeGammaReverse(i, fade_current), 
                                             0, Light.fade_duration,
                                             fadeGammaReverse(i, Light.fade_start_10[i]),
                                             fadeGammaReverse(i, Light.fade_end_10[i])));
```
the fade_current variable holds the number of milliseconds into the fade, but it is being used as argument for the fadeGammaReverse function which converts colors. the fadeGammaReverse function call should be removed and fade_current should be passed as first argument to ensure correct scaling.

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by pimw1** [comment]
Hi, thank you so much for this post. I have exactly the same issue.
The awesome smooth dimming strategy as described here <https://github.com/arendst/Tasmota/pull/11269> , by using dimmer fade, followed by a dimmer stop signal (using the " !"), is completely useless when LedTable is set to 1.
When dimming up, and stopping the dimming process, the brightness jumps back quite a bit. It's very noticable, it looks very weird. Its creating the opposite experience than what we want to achieve with "smooth dimming".

Therefore, i feel this needs more attention that it is currently given. As a work-around, i am disabling the LedTable (LedTable set to 0). By disabling the LedTable, there is no weird behaviour.

**Comment by JosefRypacek** [comment]
I'm using Shelly RGBW2 with multiple white LED strips - `Channel` command.

My setting is
```
LedTable 1
Fade 1
Speed 20
```

Brightness change between 0-30 is almost instant (< 1s) while it should take approx. 3s. The rest (30-100%) seems to be fine (slow, smooth...). There is no significant issue without LedTable.

I'm available to gather some additional date if required.

**Reply by s-hadinger** [reply]
The imbalance for Fade is on-purpose. If the Fade is too slow, switching the light on gives the impression that it is not responsive and people may try again. So the Fade from black is on-purpose accelerated. The documentation is missing this information.

Internally `gamma_table_fast[]` is used for fade from black, instead of the regular `gamma_table[]`.

I can add a compile time option to revert to the standard Gamma if you want to try, I'm not sure you will be happy with the result.

**Reply by JosefRypacek** [reply]
I think it's much better now, after #18666 (If I remember previous behavior correctly). At least it seems to be almost linear when changing 0-100 and otherwise.

I tried to replace `gamma_table_fast` with `gamma_table` and saw that it is really slow and current approach is fine.
Update of documentation would be helpful.

I did some basic measurement with `Speed 40` and changing brightness as below just to understand that.

```
100->0   20s
 80->0   15s
 60->0    8.5s
 40->0    3.5s
 20->0    0.5s
```

I also confirmed that the result is the same for all following combinations.
```
20->0
20->1
 0->20
 1->20
```
